### PR TITLE
update guard condition by managing multiple items

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
related to https://github.com/ros2/rclcpp/issues/1611

Signed-off-by: Chen Lihui <lihui.chen@sony.com>

### CODE REVIEW

File: rmw_fastrtps_shared_cpp/src/types/guard_condition.hpp [hunk 1]
```diff
@@ -80,8 +82,8 @@ class GuardCondition
 private:
   std::mutex internalMutex_;
   std::atomic_bool hasTriggered_;
-  std::mutex * conditionMutex_ RCPPUTILS_TSA_GUARDED_BY(internalMutex_);
-  std::condition_variable * conditionVariable_ RCPPUTILS_TSA_GUARDED_BY(internalMutex_);
+  std::list<std::pair<std::mutex *, std::condition_variable *>> conditions_
```
Comment (wjwwood) [inline rmw_fastrtps_shared_cpp/src/types/guard_condition.hpp:85]: I think we should make this a `std::vector`? The most common operations (in my opinion) will be `trigger()` which has to loop over this sequence of pairs. If removing a pair is expensive (inside of `detachCondition()`), I don't think that's as important.

**Comment by wjwwood** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14484/)
* Linux [Build Status](https://ci.ros2.org/job/ci_linux/14485/) (without this patch, to show failing test)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9246/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12151/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14585/)

**Comment by iuhilnehc-ynos** [comment]
> CI:
> 
> * Linux [Build Status](http://ci.ros2.org/job/ci_linux/14484/)

I have noted this `rmw_connextdds` issue at the https://github.com/ros2/rclcpp/issues/1611#issuecomment-826261608

> * Linux [Build Status](https://ci.ros2.org/job/ci_linux/14485/) (without this patch, to show failing test)

From the comments behind, I think running this CI is intent to use `https://github.com/ros2/rclcpp/pull/1640` but not `https://github.com/ros2/rmw_fastrtps/pull/527`, but there is no `repos_url` setting for the CI parameter.

> * Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9246/)

expected because of using `use_connextdds: false`

> * macOS [Build Status](http://ci.ros2.org/job/ci_osx/12151/)

`rmw_connextdds` issue as metioned before.

> * Windows [Build Status](http://ci.ros2.org/job/ci_windows/14585/)

`rmw_connextdds` issue as metioned before.

File: rmw_fastrtps_shared_cpp/src/types/guard_condition.hpp (hunk 1 above)
Comment (iuhilnehc-ynos) [reply rmw_fastrtps_shared_cpp/src/types/guard_condition.hpp:85]: Thank you, I'll update it.

**Comment by iuhilnehc-ynos** [comment]
@wjwwood 

I have made a [PR](https://github.com/ros2/rmw_connextdds/pull/51) for `rmw_connextdds`

**Comment by iuhilnehc-ynos** [comment]
not a correct fix, close it

**Pull request closed without merging**